spylib/request.py

```diff
@@ -57,34 +61,34 @@ class ServiceRequestFactory(Observable):
     def __init__(
         self,
         uuid,
-        api_key,
-        access_token=None,
+        api_key=None,
         secret=None,
         algorithm=None,
+        access_token=None,
         refresh_token=None,
         *args,
         **kwargs
     ):
         super().__init__(**kwargs)
-        try:
-            self.uuid = uuid
-            self.api_key = api_key
-            self.access_token = access_token
-            self.refresh_token = refresh_token
-            self.secret = secret
-            self.algorithm = algorithm
-            if self.access_token:
-                decode(access_token, secret, algorithms=[algorithm])
-        except ExpiredSignatureError:
+
+        # Ensure either the API or the access token must be set
+        assert bool(api_key or access_token)
+
+        self.uuid = uuid
+        self.api_key = api_key
+        self.access_token = access_token
+        self.refresh_token = refresh_token
+        self.secret = secret
+        self.algorithm = algorithm
+
+        # Attempt to ensure that the access token is usable
+        if self.access_token:
             try:
-                self.refresh_access_token(refresh_token)
-            except RefreshException:
-                self.login(uuid, api_key)
-        except (DecodeError, KeyError, Exception) as e:
-            raise e
+                decode(self.access_token, self.secret, algorithms=[self.algorithm])
+            except Exception:
+                self.fetch_new_tokens()
```

Comment:
I'm in favor of being more explicit here with exception handling. The reason being is that there is only one happy path case where we want to fetch new tokens - and that's when the JWT is expired. 
Other cases, such as a decode or key error - could happen due to bad configuration or perhaps an actor trying to use a token we haven't signed. Those kind of cases should probably raise an error and potentially even be logged to our alerting.

Reply:
Good thought! I'll check explicitly for an expired token, otherwise report the failure.